docs/react-jss.md

```diff
@@ -587,4 +587,101 @@ const Button = () => {
 
 ## TypeScript
 
-TODO hooks support
+React-JSS enables you to supply types for your `ruleNames`, `props`/`data`, and `theme` both explicitly and implicitly.
+
+### Explicit Typing
+
+```tsx
+// Define Component
+type RuleNames = 'myButton' | 'myLabel'
+
+interface ButtonProps {
+  children?: React.ReactNode
+  spacing?: number
+  fontWeight?: string
+  labelColor?: string
+}
+
+interface CustomTheme {
+  background: string
+}
+
+const useStyles = createUseStyles<RuleNames, ButtonProps, CustomTheme>({
+  myButton: {
+    padding: ({spacing}) => spacing || 10
+  },
+  myLabel: ({theme, ...props}) => ({
+    display: 'block',
+    color: props.labelColor || 'red',
+    fontWeight: props.fontWeight || 'bold',
+    backgroundColor: theme.background || 'gray'
+  })
+})
+
+function Button({children, ...props}: ButtonProps): React.ReactElement {
+  const theme = useTheme<CustomTheme>()
+  const classes = useStyles({...props, theme})
+
+  return (
+    <button className={classes.myButton}>
+      <span className={classes.myLabel}>{children}</span>
+    </button>
+  )
+}
+
+// Create App
+const theme = {background: 'black'}
+
+function App(): React.ReactElement {
+  return (
+    <ThemeProvider theme={theme}>
+      <Button fontWeight="bold">Submit</Button>
+    </ThemeProvider>
+  )
+}
+```
+
+You only need to supply as much type information as you need. For instance, if you aren't using a `Theme`, then you don't need to supply a type for it. You'll notice that if you want to supply types for `Props` or `Theme`, you'll have to supply the type for `RuleNames` first. This means that if you want autocomplete for the `classes` that come from `useStyles`, you'll need to explicitly provide the strings you'll be using. This restriction is due to a limitation in TypeScript [that will hopefully get resolved soon](https://github.com/microsoft/TypeScript/pull/26349).
+
+### Implicit Typing
+
+If you want to provide `Props` and/or `Theme` types without having to supply the `RuleNames` type, you can make TypeScript infer the types implicitly:
```

Comment:
Another option that may be worth documenting:

If `CustomTheme` is the same across your whole application, TypeScript and JavaScript users can create a file called `global.d.ts` anywhere in their project. It is not imported by anything; TypeScript and/or VSCode's TSServer finds it automatically. It could like this:

```ts
declare global {
  namespace Jss {
    export interface Theme {
      themeColour: string;
    }
  }
}

export {};
```

The benefit is that you don't need to specify `CustomTheme` when you call `createUseStyles`. There is a bit more info in #1453

Reply:
Yeah I was thinking about that, but I wasn't sure how to introduce it. What you provided gave me a good amount of inspiration though. Thanks!